Raspberry Pi cloud-init files.

wtgee opened the pull request:
These are the files are meant to be placed on a newly imaged raspberry Pi running ubuntu-server 64bit.

The instructions given by ubuntu should be followed: https://ubuntu.com/tutorials/how-to-install-ubuntu-on-your-raspberry-pi#1-overview

Section 3 of that guide talks about editing and moving the `network-config` file.  In our case we want to move all of these files.

So,

1. Install Ubuntu onto card via Imager on laptop.
2. When done, remove and reinsert SD card to laptop.
3.Change the `instance_id` in `meta-data` to be the camera name.
4. Change the `hostname` in `user-data` to be the camera name.
5. Add the NTP server at the top of the list in `user-data`.
4. Copy `user-data`, `meta-data`, `network-config`, and `usercfg.txt` to  `/system-boot` of the newly mounted SD card.
5. Unmount SD card and put in camera pi.
6. Boot pi.
7. Wash, rinse, repeat 1-7 for all cameras.
8. Profit. :pie: :money_mouth_face: 

This assumes the Pis have ethernet connection while being installed.  Wifi connection is not as simple as just editing the `network-config` if we want the other things in `user-data` to work.

In the future we could feasibly host an actual `cloud-init` instance (locally on the control computer), and then we would just have to edit one file, point it to the cloud, and it downloads the remaining files. It's more dynamic that way but not necessary for us.

Note that these are not complete yet.  We can set up all the shared ssh in here, install packages, etc. See [more examples here for ideas](https://cloudinit.readthedocs.io/en/latest/topics/examples.html).

wtgee committed:
Raspberry Pi cloud-init files.

wtgee committed:
* Remove ntp

wtgee committed:
* Updates to rpi files.

AnthonyHorton commented on resources/rpi/user-data:
```diff
@@ -0,0 +1,74 @@
+  - gh:huntsmanscope
+
+chpasswd:
+  expire: true
```
Is this enabling password expiry?

wtgee replied:
It causes it to expire immediately so you are forced to change the password on first login.  I'll add a comment.

AnthonyHorton commented on resources/rpi/user-data:
```diff
@@ -0,0 +1,74 @@
+
+ntp:
+  enabled: true
+  servers:
```
Might want to put the SSO ntp server in here, top of the list.

wtgee replied:
Actually, I did in the first commit and then got worried @lspitler wouldn't like that. :shushing_face:

AnthonyHorton replied:
Oh yeah, IPs in public repos. Not good.

AnthonyHorton commented on resources/rpi/user-data:
```diff
@@ -0,0 +1,74 @@
+    - time4.google.com
+
+## Install additional packages on first boot
+packages:
```
Do we want to include the non-free/binary SoC drivers that are mentioned in the (apparently really, really out of date) Ubuntu Raspberry Pi wiki? I know we're running headless so maybe it doesn't matter.

wtgee replied:
I think we'll probably fill this out a lot more. There's also the option below to run scripts, so we could just have the install script that changes.

AnthonyHorton commented on resources/rpi/usercfg.txt:
```diff
@@ -0,0 +1,31 @@
+# PANOPTES POCS Pi customisation for Raspberry Pi 4B+
+
+dtoverlay=disable-bt
+dtoverlay=i2c-gpio
```
Only the domepi needs GPIO, but it probably doesn't hurt to have it enabled on the others too.

AnthonyHorton reviewed (approved):
Looks good to me, at least as a starting point.

wtgee committed:
Update user-data

Adding notes about default user.

wtgee commented:
FYI, I think we will want to fix up the ssh first to use this.

wtgee commented:
FYI, this PR is editable so feel free to just make changes.  You can do it directly here in github if you don't want to checkout, etc.

wtgee commented on resources/rpi/user-data:
```diff
@@ -0,0 +1,76 @@
+# key automatically so that you can log into the unit without
+# a password. Change 'CHANGEME' to your username below.
+ssh_import_id:
+  - gh:huntsmanscope
```
@lspitler do you own the `huntsmanscope` user?  We can do make some things slightly easier with ssh keys if you have a public ssh for that account.  If not we can create one.

wtgee committed:
Update user-data

Selfish personal installs and also removing some older PANOPTES items.

wtgee committed:
Update user-data

Adding some comments for where items needed to be edited.

danjampro reviewed (commented):
@wtgee Looks good - thanks for doing this! Do you think we should install the camera and filter wheel libraries on first boot?

wtgee commented:
> @wtgee Looks good - thanks for doing this! Do you think we should install the camera and filter wheel libraries on first boot?

We want to do as much as absolutely possible in these files so that we don't have to do any manual work on the Pi once it is booted.

wtgee committed:
Update user-data

Starting to add some post install instructions.

Commit:
Moved ntp to the top of user-data

JAAlvarado-Montes commented:
Moved ntp info to the top of the file

lspitler committed:
modify camera docker-compose to a working version

lspitler commented:
On Pi8, I've setup further from the steps at top of PR using https://github.com/AstroHuntsman/huntsman-pocs/wiki/Setup-raspberry-Pi-as-a-camera-server#setting-up-a-docker-version. This is what I had to do:

```
sudo apt-get update
sudo apt-get -y install git python3-pip byobu vim grc
curl -fsSL get.docker.com -o get-docker.sh && sh get-docker.sh
apt-get install docker-compose
```

When ready to run docker:

```
docker pull huntsmanarray/camera:[github_huntsman_pocs_branch_name]
wget https://raw.githubusercontent.com/wtgee/huntsman-pocs/add-rpi-cloud-init-files/docker/camera/docker-compose.yml
```

Then just:

```
docker-compose run camera
```

Non-urgent, useful things:

- passwordless sudo

wtgee commented:
> Non-urgent, useful things:
> 
>     * passwordless sudo

This was not working?  It is set up in the `user-data` file already should already be done.

wtgee committed:
* Ensure run with root privileges.
* Fix the `ARCH` for `armv8`
* Output envvars before build.

wtgee committed:
* Slight cleanup.

wtgee committed:
* Using simplified arm device base docker based on panoptes-utils.
* Rename script.

wtgee committed:
Docker cleanup to build for some pocs usage and newer cameras.

wtgee committed:
* Whitespace.

wtgee committed:
* Added entrypoint that uses `huntsman` user.
* Consistent env var passing to docker compose.

wtgee committed:
* Update GHA docker build.
* Camera docker file default tag is `develop` but can override.

wtgee committed:
* Placeholders for ssh keys.
* Placeholders not commented out so it will break if you try to use it, forcing you to update.

wtgee committed:
* More packages.

wtgee committed:
* Write out the fuse conf and a default zshrc.

lspitler commented:
![](https://github.trello.services/images/mini-trello-icon.png) [setup and test cloud init](https://trello.com/c/wZKp2f9B/510-setup-and-test-cloud-init)

wtgee committed:
Update user-data

Remove ssh key placeholders.

fergusL committed:
Need the cloud-init shebang at top of file

fergusL committed:
Added the camera server setup steps from wiki and cronjob example

camera server setup: https://github.com/AstroHuntsman/huntsman-pocs/wiki/Setup-raspberry-Pi-as-a-camera-server

wtgee commented:
Time sync - chronicle

lspitler commented:
@wtgee @danjampro https://trello.com/c/78SKh7oi/383-put-all-devices-onto-sso-ntp-server-time

lspitler commented:
Closing as https://github.com/AstroHuntsman/huntsman-pocs/pull/246 and https://github.com/panoptes/POCS/pull/1030 supersede this PR.
